SI-5940 impls are no longer in macro def pickles

xeno-by opened the pull request:
The first officially released version of macros persisted
macro def -> impl bindings across compilation runs using a neat trick.

The right-hand side of macro definition (which contains a reference to an impl)
was typechecked and then put verbatim into an annotation on macro definition.

This solution is very simple, but unfortunately it's also lacking. If we use it
then signatures of macro defs become transitively dependent on scala-reflect.jar
(because they refer to macro impls, and macro impls refer to
scala.reflect.macros.Context defined in scala-reflect.jar).
More details can be found in https://issues.scala-lang.org/browse/SI-5940.

Therefore we have to avoid putting macro impls into binding pickles and
come up with our own serialization format. Situation is further complicated
by the fact that it's not enough to just pickle impl's class and method names,
because macro expansion needs knowledge about the shape of impl's signature
(which we can't pickle). Hence we precompute necessary stuff (e.g. the layout
of type parameters) when compiling macro defs.

xeno-by committed:
SI-5940 impls are no longer in macro def pickles

The first officially released version of macros persisted
macro def -> impl bindings across compilation runs using a neat trick.

The right-hand side of macro definition (which contains a reference to an impl)
was typechecked and then put verbatim into an annotation on macro definition.

This solution is very simple, but unfortunately it's also lacking. If we use it
then signatures of macro defs become transitively dependent on scala-reflect.jar
(because they refer to macro impls, and macro impls refer to
scala.reflect.macros.Context defined in scala-reflect.jar).
More details can be found in https://issues.scala-lang.org/browse/SI-5940.

Therefore we have to avoid putting macro impls into binding pickles and
come up with our own serialization format. Situation is further complicated
by the fact that it's not enough to just pickle impl's class and method names,
because macro expansion needs knowledge about the shape of impl's signature
(which we can't pickle). Hence we precompute necessary stuff (e.g. the layout
of type parameters) when compiling macro defs.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/151/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/856/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/151/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/856/

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/152/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/857/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/152/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/857/

xeno-by committed:
macro implementations must be public

The immediate motivator for this change was the desire to keep things simple
during macro expansion.

Here's the story. When you expand a macro, you need to reflectively load
a macro implementation method and invoke it. However private methods in Scala
can sometime have their names mangled, so the reflector has to check multiple
variants of the same name (the normal one and the mangled one).

The problem is that since the previous commit we don't have an access to a
symbol of the macro impl, hence we cannot just use `impl.expandedName` like
we did before. Sure I could duplicate the logic of expandedName, but I have
a better suggestion.

Let's prohibit non-public macro implementations. This doesn't seem to hurt,
and it lets us avoid extra bit of complexity in Macros.scala.

If this measure turns out to be a hassle during the trial period of macros,
non-public macros can always be allowed. In fact, we can even have this feature
back for free when we migrate from Java reflection to Scala reflection for
invoking macro implementations (because Scala reflection knows how to account
for mangled private names). But that's the 2.10.x business.

xeno-by commented:
PLS REBUILD ALL

xeno-by commented:
review by @adriaanm or @paulp or @odersky

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/164/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/869/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/164/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/869/

paulp commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -48,6 +49,181 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
+      var payload = Map[String, Any]("versionFormat" -> versionFormat)
+      payload += "className" -> className
+      payload += "methodName" -> macroImpl.name.toString
+      payload += "signature" -> signature
```
This is unnecessarily inefficient, imperative, and mutational.  And given that you immediately turn it into a list, just make a list.

```
val payload = List[(String, Any)](
  "versionFormat" -> versionFormat,
  "className" -> className,
  "methodName" -> macroImpl.name.toString,
  "signature" -> signature
)
```

xeno-by replied:
Thank you. Updated the code

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/187/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/892/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/187/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/892/

jsuereth commented:
Review by paulp

xeno-by commented:
Prolly you wanted to say @paulp

jsuereth commented:
I did.  I meant to say "review again".

paulp committed:
Rewrite of transformTypeTagEvidenceParams.

This fixes what a bug where the prefix seems to be checked but
is ignored; however I rewrote this method purely out of a desire
for tighter code.  I would consider it a great personal
favor if you would study the before/after and endeavor to
internalize the mechanisms I applied.

Points of particular note:

 - Before you write Option[Symbol], remember we already have
   an Option[Symbol], and it's Symbol.  None == NoSymbol.
 - flatMap can be to an Option, you don't have to create a
   one element List for flatMap's enjoyment
 - lastOption and headOption are rarely the way to go
   ... especially if you start with if (xs.isEmpty) return
 - The method which checks for emptiness is ".isEmpty",
   it is NOT ".length == 0"

I should add a pattern matcher warning for when you have a lower
case name in scope (like "contextParam") and then use a lower case
pattern variable of the same name, which of course ignores the
one outside the pattern and matches everything.

Review by @xeno-by.

paulp commented:
Sorry, reviews take me a while because I tend to run into stuff which I am genetically incapable of failing to rewrite.  #1129.

paulp commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -831,43 +929,36 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
+        argss = if (hasImplicitParams) argss.dropRight(1) :+ (tags ++ argss.last) else argss :+ tags
+
+        // nb! varargs can apply to any parameter section, not necessarily to the last one
+        for ((as, i_argss) <- argss.zipWithIndex) yield {
```
Please use mapWithIndex, which is more efficient and clearer.

paulp commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -631,103 +808,44 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
+                  macroTraceVerbose("classloader is: ")(ReflectionUtils.show(cl))
+                  val implObj = ReflectionUtils.staticSingletonInstance(cl, className)
+                  // relies on the fact that macro impls cannot be overloaded
+                  // so every methName can resolve to at maximum one method
```
In that case you can use "find", not "filter", and write the last few lines as

```
implObj.getClass.getDeclaredMethods.find(_.getName == methName) map ((implObj, _)) orElse throw new NoSuchMethodException(...)
```

paulp commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -631,103 +808,44 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
+            case binding =>
+              val className = binding.className
+              val methName = binding.methName
+              macroLogVerbose(s"resolved implementation as $className.$methName")
```
Keep in mind you can always do

```
import binding.{ className, methName }
```

(I say "keep in mind" because whether that's a good idea is situational.)

paulp commented on src/reflect/scala/reflect/internal/TreeInfo.scala:
```diff
@@ -372,6 +372,13 @@ abstract class TreeInfo {
+  /** If this tree represents a type application the type arguments. Otherwise Nil.
+   */
+  def typeArguments(tree: Tree): List[Tree] = tree match {
+    case TypeApply(_, targs) => targs
```
Should this catch AppliedTypeTree too?

xeno-by replied:
Staying consistent with firstTypeArgs declared a few lines above. For my personal use case it doesn't matter though.

paulp commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -831,43 +929,36 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
+            if (isVarArgsList(ps)) as.take(ps.length - 1) :+ as.drop(ps.length - 1)
+            else as
+          } else as
         }
```
I'd rewrite this if I could figure out what it does without too much brain work; instead I'll just ask you to comment it more.

paulp commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -831,43 +929,36 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
-        val rawArgss = for ((as, ps) <- argss zip paramss) yield {
-          if (isVarArgsList(ps)) as.take(ps.length - 1) :+ as.drop(ps.length - 1)
-          else as
+        val hasImplicitParams = macroDef.paramss.flatten.lastOption map (_.isImplicit) getOrElse false
```
```
macroDef.paramss.flatten.lastOption exists (_.isImplicit)
```

paulp commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -771,6 +889,11 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
     collectMacroArgs(expandee)
+
+    val argcDoesntMatch = macroDef.paramss.length != exprArgs.length
+    val nullaryArgsEmptyParams = exprArgs.length == 0 && macroDef.paramss.length == 1 && macroDef.paramss.flatten.length == 0
```
Please don't ever write ".length == 0".  It's "isEmpty".  It doesn't matter if you know it's short, it's just a really bad habit.

The second two conditions appear to be

```
macroDef.paramss == List(Nil)
```

paulp commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -48,6 +49,183 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
+      }.transform(pickle)
+    }
+
+    def unpickle(pickle: Tree): Option[MacroImplBinding] =
```
In general (and here specifically) you should separate the error handling from the logic.  Write a clean method which returns MacroImplBinding, and wrap a call to that method with the exception handling code (in another method if appropriate.)This makes code way easier to reuse, including within the same file -- and if there's one skill where we are in desperately short supply, it's successful code reuse.

xeno-by replied:
I'll try to figure something out. It won't be easy though, because I remember erroneous trees giving me hard time, essentially forcing to account for errors all over the file.

paulp commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -48,6 +49,183 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
+
+        val sym = macroImpl.owner
+        if (sym.isClass || sym.isModule) loop(sym)
+        else loop(sym.enclClass)
```
These three lines of code are equivalent to

```
loop(macroImpl.owner.enclClass)
```

I will leave why as an exercise to the reader. (But please ask if you can't deduce it.)

paulp commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -787,33 +910,8 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
-          typer.TyperErrorGen.MacroPartialApplicationError(expandee)
-          return None
-        }
+        val binding = loadMacroImplBinding(macroDef).get
```
Oh look, already an illustration of why I said to separate the error handling and the logic.
Anytime you find yourself doing a blind get on an Option, it's because you should have been
able to call a method which returned T, not Option[T], but you couldn't.

xeno-by committed:
accommodated the feedback for the pull request

For the discussion see: https://github.com/scala/scala/pull/1101

paulp commented:
If all my comments are addressed then this LGTM -- although I admit that my understanding of the actual macro part is not that great.  However, making the changes I've indicated (here, and everywhere) will greatly accelerate the speed with which I or anyone is likely to understand the macro-specific logic.

xeno-by commented:
Ditching `Option` in `loadMacroImplBinding` allowed to simplify even more code. Indeed I should be more careful around options.

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/195/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/901/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/195/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/901/

xeno-by commented:
These are legitimate failures. I will address them tomorrow morning.

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/197/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/903/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/197/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/903/

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/199/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/905/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/199/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/905/

xeno-by commented:
Can I has merge?
